### DISCUSSION TITLE
[CLOSED] Consider creating duckdb-nightly on PyPI separately for per commit build

### DISCUSSION THREAD

**Discussion by TechCiel** [duckdb/duckdb#5811] [General]
Hi,

I'm currently maintaining a public mirror of PyPI, and noticed the significantly high (1st place, doubled the 2nd) frequency (also the traffic) of release for `duckdb` on PyPI. I guess each commit on this mainline cause the Actions to pack a pre-release version, and then publish to PyPI.

As a prior art, the tensorflow project use packages like `tf-nightly` to publish frequently changed dev versions, while also only do packaging literally nightly. As every dev version of `duckdb` costs hundreds of megabytes to sync to mirrors over the world, for god's sake, would you please limit the packaging frequency down, or use `duckdb-nightly` for pre-releases as a way to filter?

Thanks in advance for any cooperation.

**Comment by Mause** [comment]
Hi! Just have a couple of clarifying questions.

1. Why are you maintaining an *eager* mirror of PyPI?
2. Related to the above, why not just mirror package versions and architectures that are requested, rather than all of them?
3. If DuckDB is such an issue, why not just elect to not mirror prebuilds?
4. How is breaking prebuilds out into duckdb-nightly any different from publishing prebuilds?

Cheers!

**Reply by TechCiel** [reply]
Thanks a lot for attention.

1. There's a official tool called `bandersnatch` for organizations and universities to maintain a local mirror internally (or publicly in my case). It's not something unusual.
2. Most packages does not change a lot, so will not induce heavy traffic syncing. But `duckdb` is an exception, as it make tens of ~550 MiB-ish build per week.
3. Some projects do make a good use of pre-releases, but the `bandersnatch` cannot filter out pre-releases on a per project manner (all or none). So opt-out pre-releases just for `duckdb` is not possible.
4. We can simply exclude `*-nightly*` packages, just as we do for Tensorflow nightlys. I'd not bother you if `bandersnatch` can ignore pre-releases only for `duckdb`, but it's not the case.

Cheers!

**Reply by Mause** [reply]
1. I was referring mostly to the **eager** part - that seems unusual in my experience
2. This is fair, and something we've run into with regard to package size on PyPI
3. Looks like you can filter out by release version here? https://bandersnatch.readthedocs.io/en/latest/filtering_configuration.html#regex-filtering
4. See above

**Reply by TechCiel** [reply]
1. I'm not sure what you mean by *eager*. If you mean short sync interval, that's because PyPI has a serial-based changelog allowing incremental sync with little overhead.
3. That is a global filter against the `version` string, which does not include the project name part. `bandersnatch` did not specify a way for combining multiple filters, either.

**Reply by Mause** [reply]
1. By eager I mean, ahead of the package being requested. My experience with PyPI mirrors is with Nexus (at my day job, a large enterprise), which only caches packages when users request them. This means the entire index need not be duplicated, so only versions and architectures that are requested are cached
2. It sounds like then that you might have to write a custom filter plugin, which looks to be fairly straightforward, see here: https://github.com/pypa/bandersnatch/blob/main/src/bandersnatch_filter_plugins/prerelease_name.py

**Reply by TechCiel** [reply]
1. Because this is a public mirror supported by university. Mirroring is a slightly different concept from caching.
2. Didn't see a proper way documented to load a 3rd-party filter, though some hack might be done.
2.1. As I mentioned before, `duckdb` is the only package doing this pre-release flooding in their main package. `duckdb` made 65 updates in about 10 days to be the first, 35 the second with much smaller release size. Pre-release is anyway something like *release* instead of *snapshot build*. Could this project kindly follow the nightly approach (just use cron for Actions not push!) or the `-nightly` approach (separate package) to avoid this?
2.1.1. If you think current approach is very proper and appropriate, you have the absolute right to say no under the PyPI ToS. You may just say it.

**Reply by Alex-Monahan** [reply]
The filter plugin mentioned by @Mause is likely your fastest approach to solving this. Just filter out the word `dev`. 
We publish many changes to PyPi so that DuckDB is very easy for new folks to test if their bugs are fixed after a patch. We have used this approach for multiple years now and this is the first issue of this kind.

**Reply by TechCiel** [reply]
@Alex-Monahan , that will filter out all releases with `.dev` for ALL projects on PyPI, not just for `duckdb`. Reduce the frequency to once a day is still sufficient for *eager* users, from my point of view.

**Reply by Alex-Monahan** [reply]
That would slow the feedback cycle down quite a bit in some cases. I see that you have posted similar issues in other repos. Please consider improving your architecture instead rather than trying to change multiple other repositories. Surely other mirrors have solved this problem?

**Reply by Alex-Monahan** [reply]
Maybe this example works better? It appears to filter based on both package name and release name:
https://github.com/pypa/bandersnatch/blob/main/src/bandersnatch_filter_plugins/blocklist_name.py

```PY
Check if the package name and version matches against a blocklisted
        package version specifier
```

**Reply by TechCiel** [reply]
The only tool for PyPI mirroring seems to be the `bandersnatch` by PyPA. Other mirrors just didn't do the statistics or don't mind some extra pointless traffic. AFAIK, they only see tensorflow-*-nightly's large **total size** to exclude them, but ignored the far highest update frequency of `duckdb` with *not that large* **release size**.

That is the case there is an `and`. The settings use a [pep440 version specifier](https://peps.python.org/pep-0440/#version-specifiers), which defines: 

```
Pre-releases of any kind, including developmental releases, are implicitly excluded from all version specifiers, unless...
```

But just before I thought the problem is solved, I see the line https://github.com/pypa/bandersnatch/blob/6e5d3fbbab30e63db12ca6caad9e0a1afed7d96e/src/bandersnatch_filter_plugins/blocklist_name.py#L136 explicitly included pre-releases. I didn't see any documented way to exclude pre-releases explicitly in a pep440 ver spec.

Any one of the measures will help: reduce frequency, or use separate `duckdb-nightly` or `duckdb-dev` like what tensorflow does.

I'd thank you for you patience and your attempt to help.

P.S. I'm not some tensorflow-loving AI/ML guy, but they really did a good job on this matter.

**Reply by Alex-Monahan** [reply]
Could you not make a pull request against bandersnatch, or fork it? I truly do not see why you are not taking ownership of the issues you are having with your mirror and are instead insisting that we invest time and effort on your behalf. Please invest a few hours attempting to resolve this on your side and let us know what you come up with.

**Reply by TechCiel** [reply]
Ended up adding a `packages` config for prerelease filter, PR opend for bandersnatch. Still I don't think this wild publish-per-commit policy appropriate, that's the root motivation of this issue. But as it's not prohibited, I'd better just be satisfied having my own problem solved.

I'll pronounce this thread closed. Sorry if ruined your days 😢, and wish y'all a happy new year.
